File: lib/data/vaccin_types.dart
// Types de vaccins par espèce, avec délai de rappel (années) et durée de
// validité de la première injection (jours) — partagé entre toutes les
// fiches animal (éleveur/association via animal_fiche.dart, particulier via
// animal_fiche_particulier.dart) pour ne jamais diverger entre profils.
const Map<String, List<(String label, int rappelAns, int validiteJours)>> typesVaccinParEspece = {
  'chien': [
    ('Rage', 3, 21),
    ('CHPPI (Carré, Hépatite, Parvovirose, Parainfluenza)', 1, 0),
    ('Leptospirose (L4)', 1, 0),
    ('Toux du chenil (Bordetella)', 1, 0),
    ('Piroplasmose', 1, 0),
    ('Autre', 1, 0),
  ],
  'chat': [
    ('Rage', 3, 21),
    ('Typhus (panleucopénie)', 1, 0),
    ('Coryza', 1, 0),
    ('Leucose (FeLV)', 1, 0),
    ('PIF (coronavirus félin)', 1, 0),
    ('Autre', 1, 0),
  ],
  'lapin': [
    ('Myxomatose', 1, 0),
    ('VHD (maladie hémorragique)', 1, 0),
    ('Autre', 1, 0),
  ],
  'cheval': [
    ('Rage', 3, 21),
    ('Grippe équine', 1, 0),
    ('Tétanos', 1, 0),
    ('Autre', 1, 0),
  ],
};
// Espèces hors liste ci-dessus (oiseau, nac, ovin, caprin, porcin, autre) :
// choix générique, toujours avec "Autre" pour saisie manuelle libre.
const List<(String label, int rappelAns, int validiteJours)> typesVaccinDefaut = [
  ('Rage', 3, 21),
  ('Autre', 1, 0),
];
List<(String label, int rappelAns, int validiteJours)> typesVaccinPour(String? espece) =>
    typesVaccinParEspece[espece] ?? typesVaccinDefaut;

File: website/src/lib/vaccinTypes.ts
// Types de vaccins par espèce, avec délai de rappel (années) et durée de
// validité de la première injection (jours) — partagé entre toutes les
// fiches animal du site (mes-animaux/[id] et association/animaux/[id]) pour
// ne jamais diverger entre profils. Même table que l'app Flutter
// (lib/data/vaccin_types.dart).
export const TYPES_VACCIN_PAR_ESPECE: Record<string, { label: string; rappelAns: number; validiteJours: number }[]> = {
  chien: [
    { label: 'Rage', rappelAns: 3, validiteJours: 21 },
    { label: 'CHPPI (Carré, Hépatite, Parvovirose, Parainfluenza)', rappelAns: 1, validiteJours: 0 },
    { label: 'Leptospirose (L4)', rappelAns: 1, validiteJours: 0 },
    { label: 'Toux du chenil (Bordetella)', rappelAns: 1, validiteJours: 0 },
    { label: 'Piroplasmose', rappelAns: 1, validiteJours: 0 },
    { label: 'Autre', rappelAns: 1, validiteJours: 0 },
  ],
  chat: [
    { label: 'Rage', rappelAns: 3, validiteJours: 21 },
    { label: 'Typhus (panleucopénie)', rappelAns: 1, validiteJours: 0 },
    { label: 'Coryza', rappelAns: 1, validiteJours: 0 },
    { label: 'Leucose (FeLV)', rappelAns: 1, validiteJours: 0 },
    { label: 'PIF (coronavirus félin)', rappelAns: 1, validiteJours: 0 },
    { label: 'Autre', rappelAns: 1, validiteJours: 0 },
  ],
  lapin: [
    { label: 'Myxomatose', rappelAns: 1, validiteJours: 0 },
    { label: 'VHD (maladie hémorragique)', rappelAns: 1, validiteJours: 0 },
    { label: 'Autre', rappelAns: 1, validiteJours: 0 },
  ],
  cheval: [
    { label: 'Rage', rappelAns: 3, validiteJours: 21 },
    { label: 'Grippe équine', rappelAns: 1, validiteJours: 0 },
    { label: 'Tétanos', rappelAns: 1, validiteJours: 0 },
    { label: 'Autre', rappelAns: 1, validiteJours: 0 },
  ],
};
// Espèces hors liste ci-dessus (oiseau, nac, ovin, caprin, porcin, autre) :
// choix générique, toujours avec "Autre" pour saisie manuelle libre.
export const TYPES_VACCIN_DEFAUT = [
  { label: 'Rage', rappelAns: 3, validiteJours: 21 },
  { label: 'Autre', rappelAns: 1, validiteJours: 0 },
];
export function typesVaccinPour(espece?: string) {
  return (espece ? TYPES_VACCIN_PAR_ESPECE[espece] : undefined) ?? TYPES_VACCIN_DEFAUT;
}
export function categorieOptions(espece?: string) {
  return [{ value: '', label: 'Sélectionner…' }, ...typesVaccinPour(espece).map(t => ({ value: t.label, label: t.label }))];
}

// Le délai légal (ex: rage = 21 jours) ne s'applique qu'à la toute première
// injection de ce type de vaccin pour cet animal : un rappel d'un vaccin
// déjà administré est valide dès le jour même.
export function suggestFromCategorie(espece: string | undefined, categorie: string, dateInjection: string, dejaVaccine: boolean) {
  const def = typesVaccinPour(espece).find(t => t.label === categorie);
  if (!def || !dateInjection) return null;
  const dInj = new Date(dateInjection);
  if (isNaN(dInj.getTime())) return null;
  const validite = new Date(dInj);
  validite.setDate(validite.getDate() + (dejaVaccine ? 0 : def.validiteJours));
  const rappel = new Date(dInj);
  rappel.setFullYear(rappel.getFullYear() + def.rappelAns);
  return { validite: validite.toISOString().slice(0, 10), rappel: rappel.toISOString().slice(0, 10) };
}
